### PR TITLE
feat: add the custom handlers for cosmwasm

### PR DESCRIPTION
**Pull request by gsk967**
## Description

closes: #XXXX

## Test the cosmwasm
- [ ] Build the binary from `build` (from this branch: `sai/wasm_handler` https://github.com/umee-network/umee/tree/sai/wasm_handler)
- Clone the umee-infra repo 
    - git clone REPO 
    - cd $REPO_PATH
- [ ] Copy the binary to umee-infra (./noob/umeed-builds/umeed-cosmwasm) 
     - cp BINARY ./noob/umeed-builds/umeed-cosmwasm
 
- [ ] Setup the multi node with cosmwasm
```
bash ./noob/cosmwasm/setup-cosmwasm.sh
```
- [ ] Run the sample cosmwasm (all tests should pass in the script) 
   - It will upload the cosmwasm contract and initiate the contract and it will do some transactions and queries on the contract
```bash
bash ./noob/cosmwasm/test-cw20-base-cosmwasm.sh
```
---

### Author Checklist

_All items are required. Please add a note to the item if the item is not applicable and
please add links to any relevant follow up issues._

I have...

- [ ] included the correct [type prefix](https://github.com/commitizen/conventional-commit-types/blob/v3.0.0/index.json) in the PR title
- [ ] added `!` to the type prefix if API or client breaking change
- [ ] added appropriate labels to the PR
- [ ] targeted the correct branch (see [PR Targeting](https://github.com/umee-network/umee/blob/main/CONTRIBUTING.md#pr-targeting))
- [ ] provided a link to the relevant issue or specification
- [ ] added a changelog entry to `CHANGELOG.md`
- [ ] included comments for [documenting Go code](https://blog.golang.org/godoc)
- [ ] updated the relevant documentation or specification
- [ ] reviewed "Files changed" and left comments if necessary
- [ ] confirmed all CI checks have passed

### Reviewers Checklist

_All items are required. Please add a note if the item is not applicable and please add
your handle next to the items reviewed if you only reviewed selected items._

I have...

- [ ] confirmed the correct [type prefix](https://github.com/commitizen/conventional-commit-types/blob/v3.0.0/index.json) in the PR title
- [ ] confirmed all author checklist items have been addressed
- [ ] reviewed state machine logic
- [ ] reviewed API design and naming
- [ ] reviewed documentation is accurate
- [ ] reviewed tests and test coverage
- [ ] manually tested (if applicable)

### CODE REVIEW

File: app/wasm/msg/handler_leverage.go [hunk 1]
```diff
@@ -0,0 +1,88 @@
...
+		Liquidator:  m.Liquidate.Liquidator,
+		Borrower:    m.Liquidate.Borrower,
+		Repayment:   m.Liquidate.Repayment,
+		RewardDenom: m.Liquidate.RewardDenom,
+	}
+	return s.Liquidate(ctx, req)
+}
+
+// HandleSupplyCollateralize handles the supply the assets and collateral their assets.
+func (m UmeeMsg) HandleSupplyCollateralize(
```
Comment (toteki) [inline app/wasm/msg/handler_leverage.go:82]:
```suggestion
// HandleSupplyCollateral handles the supply the assets and collateral their assets.
func (m UmeeMsg) HandleSupplyCollateral(
```

File: app/wasm/msg/types.go [hunk 2]
```diff
@@ -0,0 +1,60 @@
...
+import (
+	lvtypes "github.com/umee-network/umee/v4/x/leverage/types"
+)
+
+// AssignedMsg defines the msg to be called.
+type AssignedMsg uint16
+
+const (
+	// AssignedMsgSupply represents the call to supply coins to the capital facility.
+	AssignedMsgSupply AssignedMsg = iota + 0
```
Comment (toteki) [inline app/wasm/msg/types.go:12]: +0 or +1 here?

Comment (gsk967) [reply app/wasm/msg/types.go:12]: The value will be same as umee-cosmwasm types https://github.com/umee-network/umee-cosmwasm/blob/987e4046360f4449e239c84eff7692bdd994e9cb/packages/cw-umee-types/src/msg_leverage.rs#L8

File: app/wasm/msg/types.go [hunk 3]
```diff
@@ -0,0 +1,60 @@
...
+	Decollateralize *lvtypes.MsgDecollateralize `json:"decollateralize,omitempty"`
+	// Used to borrowing coins from the capital facility.
+	Borrow *lvtypes.MsgBorrow `json:"borrow,omitempty"`
+	// Used to repaying borrowed coins to the capital facility.
+	Repay *lvtypes.MsgRepay `json:"repay,omitempty"`
+	// Used to repaying a different user's borrowed coins
+	// to the capital facility in exchange for some of their collateral.
+	Liquidate *lvtypes.MsgLiquidate `json:"liquidate,omitempty"`
+	// Used to do supply and collateralize their assets.
+	SupplyCollateralize *lvtypes.MsgSupplyCollateral `json:"supply_collateralize,omitempty"`
```
Comment (toteki) [inline app/wasm/msg/types.go:57]:
```suggestion
	SupplyCollateral *lvtypes.MsgSupplyCollateral `json:"supply_collateral,omitempty"`
```

File: app/wasm/msg/types.go [hunk 4]
```diff
@@ -0,0 +1,60 @@
...
+	AssignedMsgBorrow
+	// AssignedMsgRepay represents the call to repaying borrowed coins to
+	// the capital facility.
+	AssignedMsgRepay
+	// AssignedMsgLiquidate represents the call to repaying a different user's
+	// borrowed coins to the capital facility in exchange for some of their
+	// collateral.
+	AssignedMsgLiquidate
+	// AssignedMsgSupplyCollateralize represents the call to supply and collateralize their assets.
+	AssignedMsgSupplyCollateralize
```
Comment (toteki) [inline app/wasm/msg/types.go:33]:
```suggestion
	// AssignedMsgSupplyCollateral represents the call to supply and collateralize their assets.
	AssignedMsgSupplyCollateral
```

File: app/wasm/msg/handler_leverage.go [hunk 5]
```diff
@@ -0,0 +1,88 @@
...
+	}
+	return s.Liquidate(ctx, req)
+}
+
+// HandleSupplyCollateral handles the supply the assets and collateral their assets.
+func (m UmeeMsg) HandleSupplyCollateral(
+	ctx context.Context,
+	s lvtypes.MsgServer,
+) (proto.Message, error) {
+	req := &lvtypes.MsgSupplyCollateral{Supplier: m.SupplyCollateralize.Supplier, Asset: m.Supply.Asset}
```
Comment (toteki) [inline app/wasm/msg/handler_leverage.go:86]:
```suggestion
	req := &lvtypes.MsgSupplyCollateral{Supplier: m.SupplyCollateral.Supplier, Asset: m.Supply.Asset}
```

File: app/wasm/msg/plugin.go [hunk 6]
```diff
@@ -0,0 +1,73 @@
...
+		_, err = smartcontractMessage.HandleCollateralize(sdkCtx, plugin.lvMsgServer)
+	case AssignedMsgDecollateralize:
+		_, err = smartcontractMessage.HandleDecollateralize(sdkCtx, plugin.lvMsgServer)
+	case AssignedMsgBorrow:
+		_, err = smartcontractMessage.HandleBorrow(sdkCtx, plugin.lvMsgServer)
+	case AssignedMsgRepay:
+		_, err = smartcontractMessage.HandleRepay(sdkCtx, plugin.lvMsgServer)
+	case AssignedMsgLiquidate:
+		_, err = smartcontractMessage.HandleLiquidate(sdkCtx, plugin.lvMsgServer)
+	case AssignedMsgSupplyCollateralize:
```
Comment (toteki) [inline app/wasm/msg/plugin.go:52]:
```suggestion
	case AssignedMsgSupplyCollateral:
```

**Review by toteki — COMMENTED** [review]
(Be sure to use the var names name consistent with message name for `MsgSupplyCollateral`)

A question:

Do these handlers carry any risk of bypassing `GetSigners` security? For example, the rule where `MsgSupply.Supplier` field must be the same address that signed it

**Comment by RafilxTenfen** [comment]
> Do these handlers carry any risk of bypassing `GetSigners` security? For example, the rule where `MsgSupply.Supplier` field must be the same address that signed it

It should not bypass the signer's rule, it is the same as sending a msg through the leverage `MsgServer`, but definitely we can double-check that with a script/test

**Review by robert-zaremba — COMMENTED** [review]
let's check if there is other way than the big ambiguous JSON. I indicated already half a year ago, but didn't a good example. Osmosis is also using this style. It works, so not blocking, but let's find out if there is no better way before merging.

**Review by robert-zaremba — APPROVED** [review]

**Pull request merged**